File: csharp/tests/Training.Module3.Tests/Application/CreateDeliveryQuoteTest.cs
using Training.Module3.Application;
using Training.Module3.Domain;
using Command = Training.Module3.Application.CreateDeliveryQuote.Command;

namespace Training.Module3.Tests.Application;

public sealed class CreateDeliveryQuoteTest
{
    [Fact]
    public void CalculatesStoresAndNotifiesAboutQuote()
    {
        List<DeliveryQuote> savedQuotes = [];
        List<DeliveryQuote> notifications = [];
        CreateDeliveryQuote useCase = new(
            Calculator(),
            new DelegatingQuoteRepository(savedQuotes.Add),
            new DelegatingQuoteNotifier(notifications.Add));

        DeliveryQuote result = useCase.Execute(new Command(
            "developer@example.com",
            ShippingMethod.Standard,
            new Parcel(3.00m)));

        Assert.Equal(17.28m, result.Price);
        Assert.Equal([result], savedQuotes);
        Assert.Equal([result], notifications);
    }

    [Fact]
    public void DoesNotNotifyWhenSavingFails()
    {
        List<DeliveryQuote> notifications = [];
        IQuoteRepository failingRepository = new DelegatingQuoteRepository(
            _ => throw new InvalidOperationException("Storage unavailable"));
        CreateDeliveryQuote useCase = new(
            Calculator(),
            failingRepository,
            new DelegatingQuoteNotifier(notifications.Add));

        Assert.Throws<InvalidOperationException>(
            () => useCase.Execute(new Command(
                "developer@example.com",
                ShippingMethod.Standard,
                new Parcel(3.00m))));
        Assert.Empty(notifications);
    }

    private static DeliveryPriceCalculator Calculator()
    {
        FuelSurcharge surcharge = new(0.08m);
        return new DeliveryPriceCalculator(
            [new StandardDeliveryPricePolicy(surcharge)]);
    }

    // C# lambdas cannot implement interfaces, so these adapters play the role
    // of Java's "savedQuotes::add" / "quote -> { throw ... }" lambdas.
    private sealed class DelegatingQuoteRepository(Action<DeliveryQuote> save) : IQuoteRepository
    {
        public void Save(DeliveryQuote quote) => save(quote);
    }

    private sealed class DelegatingQuoteNotifier(Action<DeliveryQuote> quoteCreated) : IQuoteNotifier
    {
        public void QuoteCreated(DeliveryQuote quote) => quoteCreated(quote);
    }
}

File: typescript/test/module3/application/CreateDeliveryQuote.test.ts
import { Decimal } from 'decimal.js';
import { describe, expect, it } from 'vitest';
import { IllegalStateError } from '../../../src/shared/errors.js';
import { Command, CreateDeliveryQuote } from '../../../src/module3/application/CreateDeliveryQuote.js';
import type { QuoteRepository } from '../../../src/module3/application/QuoteRepository.js';
import { DeliveryPriceCalculator } from '../../../src/module3/domain/DeliveryPriceCalculator.js';
import type { DeliveryQuote } from '../../../src/module3/domain/DeliveryQuote.js';
import { FuelSurcharge } from '../../../src/module3/domain/FuelSurcharge.js';
import { Parcel } from '../../../src/module3/domain/Parcel.js';
import { ShippingMethod } from '../../../src/module3/domain/ShippingMethod.js';
import { StandardDeliveryPricePolicy } from '../../../src/module3/domain/StandardDeliveryPricePolicy.js';

function calculator(): DeliveryPriceCalculator {
  const surcharge = new FuelSurcharge(new Decimal('0.08'));
  return new DeliveryPriceCalculator([new StandardDeliveryPricePolicy(surcharge)]);
}

describe('CreateDeliveryQuoteTest', () => {
  it('calculatesStoresAndNotifiesAboutQuote', () => {
    const savedQuotes: DeliveryQuote[] = [];
    const notifications: DeliveryQuote[] = [];
    const useCase = new CreateDeliveryQuote(
      calculator(),
      { save: (quote) => savedQuotes.push(quote) },
      { quoteCreated: (quote) => notifications.push(quote) });

    const result = useCase.execute(new Command(
      'developer@example.com',
      ShippingMethod.STANDARD,
      new Parcel(new Decimal('3.00'))));

    expect(result.price).toEqual(new Decimal('17.28'));
    expect(savedQuotes).toEqual([result]);
    expect(notifications).toEqual([result]);
  });

  it('doesNotNotifyWhenSavingFails', () => {
    const notifications: DeliveryQuote[] = [];
    const failingRepository: QuoteRepository = {
      save: () => {
        throw new IllegalStateError('Storage unavailable');
      },
    };
    const useCase = new CreateDeliveryQuote(
      calculator(),
      failingRepository,
      { quoteCreated: (quote) => notifications.push(quote) });

    expect(() => useCase.execute(new Command(
      'developer@example.com',
      ShippingMethod.STANDARD,
      new Parcel(new Decimal('3.00'))))).toThrow(IllegalStateError);
    expect(notifications).toHaveLength(0);
  });
});
